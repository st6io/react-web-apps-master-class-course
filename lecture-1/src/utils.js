import { get } from 'lodash';

export const currency = {
  usd: 'USD',
  bgn: 'BGN',
  eur: 'EUR',
};

export const EUR_TO_BGN_RATE = 1.95583;

const random = (min, max) => Math.random() * (max - min) + min;

const getCurrencies = () => {
  const usdToBgn = random(1.6564, 1.7923, 4);
  const usdToEur = usdToBgn / 1.95583;

  return {
    [currency.usd]: {
      [currency.bgn]: usdToBgn,
      [currency.eur]: usdToEur,
      [currency.usd]: 1,
    },
    [currency.bgn]: {
      [currency.usd]: 1 / usdToBgn,
      [currency.eur]: 1 / EUR_TO_BGN_RATE,
      [currency.bgn]: 1,
    },
    [currency.eur]: {
      [currency.usd]: 1 / usdToEur,
      [currency.bgn]: EUR_TO_BGN_RATE,
      [currency.eur]: 1,
    },
  };
};

const currentCurrencies = getCurrencies();

export const toFixed = (value, precision = 2) => {
  const f = 10 ** precision;

  return Math.round(value * f) / f;
};

export const convertCurrency = (
  value,
  currency,
  targetCurrency,
  precision = 2,
) => {
  const input = parseFloat(value);
  if (Number.isNaN(input)) {
    return '';
  }

  const ratio = get(currentCurrencies, `${currency}.${targetCurrency}`, NaN);
  if (Number.isNaN(ratio)) {
    throw new Error(`cannot convert ${currency} to ${targetCurrency}`);
  }

  const converted = value * ratio;

  return toFixed(converted, precision);
};

const delay = (result, timeout = 250) =>
  new Promise(resolve => setTimeout(() => resolve(result), timeout));

export const fetchTempertature = () => delay('20C', 1000);

export const fetchHumidity = () => delay('54%');
