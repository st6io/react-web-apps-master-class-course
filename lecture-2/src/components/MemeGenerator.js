import React from 'react';

import Container from './Container';
import Card from './Card';
import styles from './MemeGenerator.module.scss';
import MemeInput from './MemeInput';
import Meme from './Meme';

const defaultImage =
  'https://miro.medium.com/max/1200/1*y6C4nSvy2Woe0m7bWEn4BA.png';

export default class MemeGenerator extends React.Component {
  state = {
    topText: 'React',
    bottomText: 'is awesome!',
    url: '',
    image: defaultImage,
  };

  handleInputChange = input => {
    this.setState(input);
  };

  handleImageChange = image => {
    this.setState({ image });
  };

  render() {
    const { topText, bottomText, url, image } = this.state;

    return (
      <Container className={styles.container}>
        <Card className={styles.card}>
          <MemeInput
            {...{ topText, bottomText, url }}
            onInputChange={this.handleInputChange}
            onImageChange={this.handleImageChange}
          />
          <Meme {...{ topText, bottomText, image }} />
        </Card>
      </Container>
    );
  }
}
