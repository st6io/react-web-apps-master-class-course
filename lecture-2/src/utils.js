const images = [
  'https://www.maxim.com/.image/t_share/MTYwNjI1MDQ4NTM1NzA0NzQ3/jocko-willink-attack-mondays.png',
  'https://i.ytimg.com/vi/ljqra3BcqWM/maxresdefault.jpg',
  'https://peterattiamd.com/wp-content/uploads/2019/06/jocko-part2-2x1.png',
  'https://i.ytimg.com/vi/Ls5yHTGVzgU/maxresdefault.jpg',
  'https://miro.medium.com/max/3600/0*uvDOX4ea2mGXdAQ9',
];

export const getRandomImage = async () => {
  const response = await fetch(
    'https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint16',
  );

  const {
    data: [rnd],
  } = await response.json();

  return images[rnd % images.length];
};

export const sum = (a, b) => a + b;
