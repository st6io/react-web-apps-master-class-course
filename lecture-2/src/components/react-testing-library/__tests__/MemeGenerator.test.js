import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import { getRandomImage } from '../../../utils';

import MemeGenerator from '../../MemeGenerator';

jest.mock('../../../utils', () => ({
  getRandomImage: jest.fn(),
}));

describe('MemeGenerator Component', () => {
  beforeEach(() => {
    getRandomImage.mockClear();
  });

  it('should match snapshot', () => {
    const { container } = render(<MemeGenerator />);
    expect(container).toMatchSnapshot();
  });

  it('test shuffle integration', async () => {
    const { getByText, container } = render(<MemeGenerator />);
    const expectedImage = 'https://test.img/jpg';

    getRandomImage.mockResolvedValueOnce(expectedImage);

    fireEvent.click(getByText('Shuffle'));

    await wait(getRandomImage);

    expect(container.getElementsByTagName('img')[0]).toHaveAttribute(
      'src',
      expectedImage,
    );
  });
});
