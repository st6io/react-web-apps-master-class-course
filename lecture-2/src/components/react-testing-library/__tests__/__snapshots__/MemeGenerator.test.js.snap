// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemeGenerator Component should match snapshot 1`] = `
<div>
  <div
    class="container container"
  >
    <div
      class="card card"
    >
      <div>
        <div
          class="row"
        >
          <input
            name="topText"
            type="text"
            value="React"
          />
        </div>
        <div
          class="row"
        >
          <input
            name="bottomText"
            type="text"
            value="is awesome!"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="leftColumn"
          >
            <button>
              Via Url
            </button>
            <button
              id="shuffle"
            >
              Shuffle
            </button>
          </div>
        </div>
      </div>
      <div
        class="meme"
      >
        <div
          class="topText"
        >
          React
        </div>
        <div
          class="bottomText"
        >
          is awesome!
        </div>
        <img
          alt="meme"
          class="memeImage"
          src="https://miro.medium.com/max/1200/1*y6C4nSvy2Woe0m7bWEn4BA.png"
        />
      </div>
    </div>
  </div>
</div>
`;
